### PR TITLE
Combine yarpc.Meta and context.Context

### LINKED ISSUE #23: Combine Context & Meta
Let's try this out:

```
func (h Handler) Get(context Context, request Request) response, context, err
```

Will look more like node, python, & java.

### PR DESCRIPTION
**Pull request by abhinav**
This combines the `yarpc.Meta` and `context.Context` objects into a `{json, thrift}.Request` and changes use of `yarpc.Meta` in responses to a `{json, thrift}.Response`. We'll want to put response context there.

Resolves #23.

### CODE REVIEW

**Comment by abhinav** [comment]
@yarpc/golang @yarpc/yarpc

**Comment by breerly** [comment]
Let's sit on this till after the demo - I want time to check this branch out and use it for a bit.

**Comment by breerly** [comment]
I'm surprised how much smoother this api looks and feels - big :shipit: here!

File: encoding/json/request.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Context context.Context
+
+	// Name of the procedure being called.
+	Procedure string
+
+	// Request headers
+	Headers transport.Headers
+
+	// TTL is the amount of time in which this request is expected to finish.
+	TTL time.Duration
```
Comment (prashantv) [inline encoding/json/request.go]: how does `TTL` interact with the deadline stored in the `context.Context`?

File: encoding/thrift/outbound.go [hunk 2]
```diff
@@ -24,22 +24,25 @@ import (
...
-	Call(ctx context.Context, r *Request) (*Response, yarpc.Meta, error)
+	// Call the given Thrift method.
+	Call(method string, req *Request, body wire.Value) (wire.Value, *Response, error)
 }
 
 // Config contains the configuration for the Client.
 type Config struct {
+	// Name of the Thrift service. This is the name used in the Thrift file
+	// with the 'service' keyword.
+	Service string
```
Comment (prashantv) [inline encoding/thrift/outbound.go]: should we be explicit and call this `ThriftService` till we figure out common terminology?

File: encoding/json/request.go (hunk 1 above)
Comment (abhinav) [inline encoding/json/request.go]: Discussed offline. If given, we'll use the user-provided timeout to clamp down on the deadline for the whole (multi-hop) request in a different diff.

File: encoding/thrift/outbound.go (hunk 2 above)
Comment (abhinav) [inline encoding/thrift/outbound.go]: I think that's not necessary in contexts where there is no ambiguity. `thrift.Config{Service: "MyService"}` is pretty clear about which service you're talking about.

Besides, this isn't really a user-facing API. It'll be used by the generated code.

File: encoding/json/inbound_test.go [hunk 3]
```diff
@@ -76,6 +76,27 @@ func TestHandleMapSuccess(t *testing.T) {
...
+	}
+
+	resw := new(transporttest.FakeResponseWriter)
+	err := handler.Handle(context.Background(), &transport.Request{
+		Procedure: "simpleCall",
+		Body:      jsonBody(`{"name": "foo", "attributes": {"bar": 42}}`),
+	}, resw)
+	require.NoError(t, err)
+
+	assert.Equal(t, transport.Headers{"foo": "bar"}, resw.Headers)
```
Comment (prashantv) [inline encoding/json/inbound_test.go]: optional: should we retype `transport.Headers` so that a user doesn't need to interact with that package in any way.
and from an efficiency perspective, we would just cast it since they're the same.

File: encoding/json/request.go (hunk 1 above)
Comment (prashantv) [inline encoding/json/request.go]: Cool, let's rename this and update the comment to note that it's optional and if specified, will only ever clamp down the context's deadline.

File: encoding/thrift/outbound.go [hunk 4]
```diff
@@ -80,70 +85,62 @@ type thriftClient struct {
...
-	// 		resp, m, err := c.client.Call(ctx, &thrift.Request{
-	// 			Service: "MyService",
-	// 			Method: "someMethod",
-	// 			Meta: m,
-	// 			Body: args.ToWire(),
-	// 		})
-	// 		if err != nil { return nil, m, err }
-	// 		if resp.Exception1 != nil {
-	// 			return nil, m, resp.Exception1
+	// 		resBoy, res, err := c.client.Call("someMethod", req, args.ToWire())
```
Comment (prashantv) [inline encoding/thrift/outbound.go]: typo: resBoy -> resBody

**Comment by prashantv** [comment]
lgtm

File: encoding/thrift/outbound.go (hunk 4 above)
Comment (abhinav) [inline encoding/thrift/outbound.go]: Fixed

File: encoding/json/inbound_test.go (hunk 3 above)
Comment (abhinav) [inline encoding/json/inbound_test.go]: Made sense. Noted. Will do in a different PR.
